webrender/src/device/gfx/program.rs

```diff
@@ -98,63 +100,78 @@ impl<B: hal::Backend> Program<B> {
 
         let (vs_module, fs_module) = shader_modules.get(shader_name).unwrap();
 
-        let mut specialization_data =
-            vec![0; (SPECIALIZATION_CONSTANT_COUNT - 1) * SPECIALIZATION_CONSTANT_SIZE];
-        let mut constants = SPECIALIZATION_FEATURES
-            .iter()
-            .zip(specialization_data.chunks_mut(SPECIALIZATION_CONSTANT_SIZE))
+        let mut specialization_data = vec![];
+        let program_modes = if let ShaderKind::Text = shader_kind {
+            [0, 3, 4, 5].into_iter()
+        } else {
+            [0].into_iter()
+        };
+        let constant_setups: ArrayVec<[ArrayVec<[SpecializationConstant; SPECIALIZATION_CONSTANT_COUNT]>; 4]> =
+            program_modes
             .enumerate()
-            .map(|(i, (feature, out_data))| {
-                out_data[0] = features.contains(feature) as u8;
-                hal::pso::SpecializationConstant {
-                    id: i as _,
-                    range: (SPECIALIZATION_CONSTANT_SIZE * i) as _
-                        ..(SPECIALIZATION_CONSTANT_SIZE * (i + 1)) as _,
+            .map(|(i, mode)| {
+                specialization_data.push(vec![0; (SPECIALIZATION_CONSTANT_COUNT - 1) * SPECIALIZATION_CONSTANT_SIZE]);
+                let mut constants: ArrayVec<[SpecializationConstant; SPECIALIZATION_CONSTANT_COUNT]> = SPECIALIZATION_FEATURES
```

Comment:
nit: you can omit the type of a element here, i.e . `ArrayVec<[_; SPECIALIZATION_CONSTANT_COUNT]>`